### DISCUSSION TITLE
Create a Linux VM from an existing raw (EFI) image

### DISCUSSION THREAD

**Discussion by dhess** [cirruslabs/tart#387] [Ideas]
Hi, thanks for this project! I've been looking for something like this that's CI-oriented, rather than focused on interactive use, for awhile.

I'd like to use Tart to run NixOS VMs. Unlike the case with most other Linux distros, it's relatively easy to create a custom NixOS disk image in a variety of formats. In particular, https://github.com/nix-community/nixos-generators makes it just a few extra lines of configuration on top of an existing NixOS config, and then you can create, say, a raw EFI image configured precisely like you want, like so:

```
nixos-generate -f raw-efi -c configuration.nix
```

From what I can gather (e.g., https://github.com/cirruslabs/tart/discussions/171#discussioncomment-3322819), `Virtualization.Framework` expects a Linux VM's disk image to be just a raw disk image, and supports EFI layouts, so in theory, it should be possible to use `nixos-generate` to create NixOS VM images that are ready to use with Tart in a single step, obviating the need to use Packer or a manual installation process from an ISO.

As an experiment, I built a minimal raw EFI image using `nixos-generate`. In my particular case, it came out to about 4.6GiB, and the `file` command on macOS says the following about it:

```
 file disk.img
disk.img: DOS/MBR boot sector; partition 1 : ID=0xee, start-CHS (0x0,0,2), end-CHS (0x3ff,255,63), startsector 1, 9508863 sectors, extended partition table (last)
```

I then created a new VM using `tart create --linux nixos`, and replaced the empty disk image that Tart created with the NixOS image I created using `nixos-generate`. Unfortunately, when I try to run the VM, I get the following error:

```
 tart run nixos                                                                                                                                                                ~
Error Domain=VZErrorDomain Code=2 "A storage device configuration is invalid." UserInfo={NSLocalizedFailure=Invalid virtual machine configuration., NSLocalizedFailureReason=A storage device configuration is invalid., NSUnderlyingError=0x600003aaabe0 {Error Domain=VZErrorDomain Code=2 "The storage device attachment is invalid." UserInfo={NSLocalizedFailure=Invalid virtual machine configuration., NSLocalizedFailureReason=The storage device attachment is invalid.}}}
```

I don't know much about `Virtualization.Framework`, so I'm flying blind here, but my first hunch is that perhaps the `nvram.bin` file created by Tart contains the expected disk size, which doesn't match the size of the image that `nixos-generate` created. To test this theory, I tried to configure my NixOS image to be precisely the same size as the image that Tart created, but unfortunately, I ran into a bug in the NixOS raw image generation that prevents me from doing this.

Before I look into that, I wanted to check with the Tart team to see whether I'm barking up the wrong tree. Perhaps there's another explanation for this error? Is there any reason to expect that this wouldn't work? I think that, irrespective of NixOS, it would be a nice feature to be able to hand Tart an existing raw (EFI) image, without depending on a specific packaging tool like Packer.

**Comment by edigaryev** [comment]
>I don't know much about Virtualization.Framework, so I'm flying blind here, but my first hunch is that perhaps the nvram.bin file created by Tart contains the expected disk size, which doesn't match the size of the image that nixos-generate created.

It shouldn't contain that information, at least not before you `tart run` the VM, because that NVRAM file [is created separately](https://github.com/cirruslabs/tart/blob/31635e59d7ca808128c7a2a8fa639b02b52cd760/Sources/tart/VM.swift#L207) from the disk.

>Perhaps there's another explanation for this error?

I'd suggest providing some concrete steps that reproduce your error first. To me it looks like you might be supplying the same disk image to the `tart run` twice, which causes the invalid VM configuration.

>I think that, irrespective of NixOS, it would be a nice feature to be able to hand Tart an existing raw (EFI) image, without depending on a specific packaging tool like Packer.

Looks like a valid use-case, you might consider [creating an issue](https://github.com/cirruslabs/tart/issues) for this.

**Reply by dhess** [reply]
> > I don't know much about Virtualization.Framework, so I'm flying blind here, but my first hunch is that perhaps the nvram.bin file created by Tart contains the expected disk size, which doesn't match the size of the image that nixos-generate created.
> 
> It shouldn't contain that information, at least not before you `tart run` the VM, because that NVRAM file [is created separately](https://github.com/cirruslabs/tart/blob/31635e59d7ca808128c7a2a8fa639b02b52cd760/Sources/tart/VM.swift#L207) from the disk.

Ahh, thank you! I think you are correct. I tried the process again, this time without trying to `tart run` the newly-created Linux VM first, and now it works!

> Looks like a valid use-case, you might consider [creating an issue](https://github.com/cirruslabs/tart/issues) for this.

OK, I'll keeping prototyping this a bit and then maybe open an issue or PR.
